### PR TITLE
[master] AbstractMotor and related classes

### PR DESCRIPTION
**Pull request by IvarsKarpics**
* AbstractMotor: Changed class type to HardwareObject to ensure that the signals are emited.
  def set_position emits positionChanged
  def set_state emits stateChanged
  def set_limits emits limitsChanged
* HardwareObjects/ExporterMotor.py: new
* HardwareObjects/ExporterZoom.py: new

Introduced new ExporterMotor and ExporterZoom because exporter channels are hard coded.

### CODE REVIEW

File: HardwareObjects/AbstractMotor.py [hunk 1]
```diff
@@ -118,6 +136,7 @@ def set_state(self, state):
             state (str): motor state
         """
         self.__state = state
+        self.emit('stateChanged', (state, ))
```
Comment (rhfogh) [inline HardwareObjects/AbstractMotor.py]: Probably just me - but why emit stateChanged and positionChanged when you are *getting* the value? Would it not be more natural to emit when you are *setting* the value?

Comment (IvarsKarpics) [reply HardwareObjects/AbstractMotor.py]: In fact the signals are emitted in set_* methods and if you look at the all code it is fine. I do not know why it is displayed here wrong.

**Comment by rhfogh** [comment]
AbstractMotor.__init__ now takes a mandatory 'name' parameter. It looks to me like this breaks MotorMockup, and in general any subtype of AbstractMotor where there is a call to this __init__ that has not been fixed.

**Comment by rhfogh** [comment]
I have now managed to get the program to start up in mock mode without errors, using the motors branch of HardwareRepository combined with the up-to-date master of mxcube. Maybe this was too early and I should not have tried yet. 
In  order to get there I had to do the following temporary fixes - which may indicate points to check in the code:

- Changed the MotorMockup class definition to start like this:
class MotorMockup(AbstractMotor):
    def __init__(self, name):
        AbstractMotor.__init__(self, name)
        self.__move_task = None

- Put the body of Qt4_MotorPredefPosBrick.motor_state_changed() inside a 
try: ... except: AttributeError block, since self.motor_hwobj.motor_states raised AttributeError

- Comment out the following line in Qt4_queue_item, as XRayImaging did not appear in queue_model_objects:
     queue_model_objects.XrayImaging: XrayImagingQueueItem,

- Comment out the last two lines of Qt4TaskToolBoxBrick.propertyChanged()
since self.task_tool_box_widget did not have any attribute 'enable_compression'

- Add a get_limits() function to ResolutionMockup, returning a dummy value

Possibly I should have combined a different set of branches. Possibly I should have waited till all the PRs were merged. Or, possibly, I am the first person to test these changes in mock mode?

**Comment by IvarsKarpics** [comment]
Hi @rhfogh ,

Thanks for a deep review. Not possibly, but for sure you are the first one who is testing these changes. I will add fixes for the AbstractMotor and MotorMockup.
Regarding other bugs... hmm.. in fact it is difficult to test a single PR, because we agreed that PR should be as small as possible. In this case I have to submit several small PRs or one big to push all changes. If it is fine I would suggest to approve this PR and I will prepare small PRs that will fix your mentioned issues.

**Comment by rhfogh** [comment]
I think approving this PR and doing fixes afterwards is a good idea, but as I do not yet have write permission to HardwareRepository I cannot do it myself.

It is indeed highly tricky to divide a big set of changes into small independent PRs (I have been there myself). Not sure if it is a better policy to make sure PRs are individually testable or to make sure they are small, in the cases where you cannot do both?

Two things one could do:
- Set up a docker with the mock objects and quickly check that it opens without errors before you make the PR. There is one in the mx3docker repository. Unfortunately I cannot do the equivalent and set up an active beamline to test on ( ;-) ).
- When changing function signatures or class inits make a global search for all places where the  function/class is called and make sure the change is compatible.

**Comment by IvarsKarpics** [comment]
Hi, I added you to the contributors list. Now you should be able to accept PR.

**Comment by rhfogh** [comment]
Hi,

Not quite sure if I should approve this PR  yet. Looking through the code, I find the line
 AbstractMotor.__init__(self) 
not just in MotorMockup, but also in:
ExporterMotor
MD2Motor
MicrodiffLight
MicrodiffMotor
Resolution
SardanaMotor
EMBL/TINEMotor
Might it not cause breakages in other places, if I approve the PR before those are fixed?

**Comment by IvarsKarpics** [comment]
True... I will prepare fixes

**Comment by rhfogh** [comment]
Hi Ivars,

OK, I am running my tests now on

mxcube: updated_widgets merged with master, with some gphl-specific local changes

HardwareRepository: motors, merged with local, gphl-specific changes

The only outstanding PRs I am missing (I believe) are HardwareRepository EMBL updates, and HardwareRepository modified queue enumerables. As far as I can tell neither of those should have any effects on the problems reported below (neither should my local changes, which do not touch the relevant files), so this ought to be a fair test (I could be wrong, of course).

I get the following errors, which go away after the temporary fixes mentioned. Could you have a look and either add a fix or maybe tell me what I am doing wrong?

Yours,

Rasmus

-  File "/MXCuBE/mxcube/BlissFramework/Bricks/Qt4_ResolutionBrick.py", line 432, in get_resolution_limits
    self.resolution_limits_changed(self.resolution_hwobj.get_limits())
  File "/MXCuBE/mxcube/HardwareRepository/BaseHardwareObjects.py", line 345, in __getattr__
    raise AttributeError(attr)
AttributeError: get_limits
FIX: Add ResolutionMockup.get_limits(self): return (0.5, 10.0)

-   File "/MXCuBE/mxcube/BlissFramework/Bricks/Qt4_queue_item.py", line 427, in <module>
    queue_model_objects.XrayImaging: XrayImagingQueueItem,
AttributeError: 'module' object has no attribute 'XrayImaging'
FIX: Comment out the line

-   File "/MXCuBE/mxcube/BlissFramework/Bricks/Qt4_MotorPredefPosBrick.py", line 122, in motor_state_changed
    s = state in (self.motor_hwobj.motor_states.READY,
  File "/MXCuBE/mxcube/HardwareRepository/BaseHardwareObjects.py", line 345, in __getattr__
    raise AttributeError(attr)
AttributeError: motor_states
FIX (temporary): put body of Qt4_MotorPredefPosBrick.motor_state_changed inside
try: [...] except AttributeError: pass

-   File "/MXCuBE/mxcube/BlissFramework/Bricks/widgets/Qt4_data_path_widget.py", line 93, in __init__
    self.data_path_layout.compression_cbox, bool, None)
  File "/MXCuBE/mxcube/BlissFramework/Bricks/widgets/Qt4_widget_utils.py", line 211, in bind_value_update
    widget.setChecked(bool(getattr(self.__model, field_name)))
AttributeError: 'PathTemplate' object has no attribute 'compression'
FIX: Added class attribute compression=False to PathTemplate

-     from widgets.Qt4_create_xray_imaging_widget import CreateXrayImagingWidget
  File "/usr/lib/python2.7/dist-packages/gevent/builtins.py", line 93, in __import__
    result = _import(*args, **kwargs)
ImportError: No module named Qt4_create_xray_imaging_widget
FIX: None - too much to change, hopefully program will open anyway

-   File "/MXCuBE/mxcube/BlissFramework/Qt4_GUISupervisor.py", line 452, in finish_init
    main_widget = self.load_gui(gui_config_file)
  File "/MXCuBE/mxcube/BlissFramework/Qt4_GUISupervisor.py", line 265, in load_gui
    main_window = self.execute(self.configuration)
  File "/MXCuBE/mxcube/BlissFramework/Qt4_GUISupervisor.py", line 366, in execute
    make_connections(config.windows_list)
  File "/MXCuBE/mxcube/BlissFramework/Qt4_GUISupervisor.py", line 363, in make_connections
    make_connections(item["children"])
[...]
  File "/MXCuBE/mxcube/BlissFramework/Qt4_GUISupervisor.py", line 363, in make_connections
    make_connections(item["children"])
  File "/MXCuBE/mxcube/BlissFramework/Qt4_GUISupervisor.py", line 359, in make_connections
    getattr(sender, connection["signal"]).connect(slot)
AttributeError: 'NullBrick' object has no attribute 'request_tree_brick'
FIX: None, program crashes.

**Comment by rhfogh** [comment]
It is indeed difficult, I know, with big changes with many ramifications. Personally I'd say that as long as whatever is in the pull requests is consistent and works, it should not matter too much how much is in the PRs. 
But this is the master branch, after all.  I do think that before making major PRs one should make a test branch that reflects what is being 'checked in' (and does not have any extraneous files or changes not part of the PRs), and check that the program loads (even better, runs), preferably both in live and mock mode.  One solution might be to do major additions (like GPhL workflows, or Xray imaging) in a separate feature branch, and move them to master once they are complete. And/or keeping commits for different purposes separate and cherry-picking across a consistent set.

**Comment by IvarsKarpics** [comment]
I fixed bugs with a new PR mxcube/mxcube#220. Changes regarding XrayImaging will be pushed as a tested feature branch.

**Comment by rhfogh** [comment]
Hi Ivars,

That took care of several of the problems. I am still seeing:

- ImportError: No module named Qt4_create_xray_imaging_widget

- AttributeError: 'PathTemplate' object has no attribute 'compression'

and

- AttributeError: 'NullBrick' object has no attribute 'request_tree_brick'

**Comment by IvarsKarpics** [comment]
Can you check now?

**Comment by rhfogh** [comment]
This time it works and the program runs - as far as starting our simulated data collection - without problems.

**Pull request merged**